src/librustc/middle/ty.rs

```diff
@@ -2301,6 +2289,128 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
     }
 }
 
+pub struct DatatypeDef<'tcx> {
+    pub def_id: DefId,
+    pub variants: Vec<VariantDef<'tcx>>,
+}
+
+pub struct VariantDef<'tcx> {
+    pub id: DefId,
+    pub name: Name,
+    pub disr_val: Disr,
+    pub fields: Vec<FieldTy<'tcx>>,
+    pub vis: Visibility,
+}
+
+
+pub struct FieldTy<'tcx> {
+    pub id: DefId,
+    pub name: Name,
+    pub vis: ast::Visibility,
+    pub origin: DefId,
+    pub ty: Cell<Ty<'tcx>>
+}
+
+impl<'tcx> DatatypeDef<'tcx> {
+
+    pub fn is_tuple_variant(&self, variant_idx: usize) -> bool {
+        if self.variants.len() <= variant_idx { return false; }
+
+        self.variants[variant_idx].is_tuple_variant()
+    }
+
+    pub fn get_variant(&'tcx self, def: DefId) -> Option<&'tcx VariantDef<'tcx>> {
+        self.variants.iter().find(|v| v.id == def)
+    }
+
+    pub fn is_univariant(&self) -> bool {
+        self.variants.len() == 1
+    }
+
+    pub fn is_struct_def(&self) -> bool {
+        self.is_univariant() && self.def_id == self.variants[0].id
+    }
+
+    pub fn is_c_like(&self) -> bool {
+        if self.variants.len() == 0 {
+            return false;
+        }
+
+        self.variants.iter().all(|v| v.fields.len() == 0)
+    }
+}
+
+impl<'tcx> VariantDef<'tcx> {
+    pub fn subst_fields<'a>(&'a self, cx: &'a ctxt<'tcx>,
+                            substs: &'tcx Substs<'tcx>) -> FieldTyIter<'a, 'tcx> {
+        FieldTyIter {
+            cx: cx,
+            substs: substs,
+            iter: self.fields.iter()
+        }
+    }
+
+    pub fn is_tuple_variant(&self) -> bool {
+        !self.fields.is_empty() &&
+            self.fields.iter().all(|f| f.name == token::special_names::unnamed_field)
+    }
+}
+
+pub struct FieldTyIter<'a, 'tcx:'a> {
+    cx: &'a ctxt<'tcx>,
+    substs: &'tcx Substs<'tcx>,
+    iter: slice::Iter<'a, FieldTy<'tcx>>
+}
+
+impl<'a, 'tcx> Iterator for FieldTyIter<'a, 'tcx> {
+    type Item = Ty<'tcx>;
+
+    fn next(&mut self) -> Option<Ty<'tcx>> {
+        self.iter.next().map(|f| f.ty.get().subst(self.cx, self.substs))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+
+impl<'tcx> PartialEq for DatatypeDef<'tcx> {
+    fn eq(&self, other: &DatatypeDef) -> bool {
+        self.def_id == other.def_id
+    }
+}
+
+impl<'tcx> Eq for DatatypeDef<'tcx> { }
+
+#[cfg(stage0)]
+impl<'tcx, S:Writer+Hasher> Hash<S> for DatatypeDef<'tcx> {
+    fn hash(&self, s: &mut S) {
+        self.def_id.hash(s);
+    }
+}
+
+#[cfg(not(stage0))]
+impl<'tcx> Hash for DatatypeDef<'tcx> {
+    fn hash<H: Hasher>(&self, s: &mut H) {
+        self.def_id.hash(s);
+    }
+}
+
+impl<'tcx> fmt::Debug for DatatypeDef<'tcx> {
```

Comment:
I added the `Debug` impl because `sty` derives `Debug` meaning that `DatatypeDef` need to impl `Debug` too.